File: backend/src/notify/dingtalk.ts
import crypto from 'node:crypto';
import { fetchWithTimeout, errorMessage } from '../util.js';
import type { DigestContent } from './format.js';

export interface SendResult {
  ok: boolean;
  error: string | null;
}

// DingTalk custom-robot signing: append timestamp + HMAC-SHA256(sign) when a secret is set.
function signedUrl(webhook: string, secret: string): string {
  const timestamp = Date.now();
  const stringToSign = `${timestamp}\n${secret}`;
  const sign = crypto.createHmac('sha256', secret).update(stringToSign).digest('base64');
  const sep = webhook.includes('?') ? '&' : '?';
  return `${webhook}${sep}timestamp=${timestamp}&sign=${encodeURIComponent(sign)}`;
}

export async function sendDingtalk(
  config: { webhook: string; secret: string | null },
  digest: DigestContent,
): Promise<SendResult> {
  try {
    const url = config.secret ? signedUrl(config.webhook, config.secret) : config.webhook;
    const res = await fetchWithTimeout(
      url,
      {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          msgtype: 'markdown',
          markdown: { title: digest.title, text: digest.markdown },
        }),
      },
      15_000,
    );
    if (!res.ok) return { ok: false, error: `HTTP ${res.status}` };
    // DingTalk returns { errcode, errmsg }; errcode 0 means success.
    const data = (await res.json()) as { errcode?: number; errmsg?: string };
    if (data.errcode && data.errcode !== 0) {
      return { ok: false, error: `dingtalk errcode ${data.errcode}: ${data.errmsg ?? ''}` };
    }
    return { ok: true, error: null };
  } catch (err) {
    return { ok: false, error: errorMessage(err) };
  }
}

File: security-intel-hub/security_intel_hub/notifiers/dingtalk.py
from __future__ import annotations

import base64
import hashlib
import hmac
import time
import urllib.parse

from security_intel_hub.http import request_json


class DingTalkNotifier:
    name = "dingtalk"

    def __init__(self, *, webhook: str, secret: str = "") -> None:
        self.webhook = webhook
        self.secret = secret

    def notify(self, title: str, text: str) -> None:
        if not self.webhook:
            return

        request_json(
            "POST",
            self._signed_webhook(),
            json_body={
                "msgtype": "markdown",
                "markdown": {
                    "title": title,
                    "text": f"### {title}\n\n{text}",
                },
            },
        )

    def _signed_webhook(self) -> str:
        if not self.secret:
            return self.webhook

        timestamp = str(round(time.time() * 1000))
        string_to_sign = f"{timestamp}\n{self.secret}".encode("utf-8")
        digest = hmac.new(self.secret.encode("utf-8"), string_to_sign, digestmod=hashlib.sha256).digest()
        sign = urllib.parse.quote_plus(base64.b64encode(digest))
        separator = "&" if "?" in self.webhook else "?"
        return f"{self.webhook}{separator}timestamp={timestamp}&sign={sign}"
